### PR TITLE
HPCC-31766 Prevent eclcc from being terminated while git is updating a repository

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/eclcc/eclcc.cpp [hunk 1]
```diff
@@ -126,16 +128,54 @@ void __cdecl IntHandler(int)
...
+        _exit(2);
+    }
+}
+
+static void installSignalHandlers()
+{
+    struct sigaction act;
+    act.sa_sigaction = &sighandler;
+    sigemptyset(&act.sa_mask);
+    act.sa_flags = SA_SIGINFO;
```
Comment (mckellyln) [inline ecl/eclcc/eclcc.cpp:164]: See above and I would probably add SA_RESTART if you really wanted to continue.

File: ecl/eclcc/eclcc.cpp [hunk 2]
```diff
@@ -126,16 +128,54 @@ void __cdecl IntHandler(int)
...
+    sigaction(SIGTERM, &act, nullptr);
+    sigaction(SIGINT, &act, nullptr);
+}
+
+static void restoreSignalHandlers()
+{
+    struct sigaction act;
+    act.sa_handler = SIG_DFL;
+    sigemptyset(&act.sa_mask);
+    act.sa_flags = SA_SIGINFO;
```
Comment (mckellyln) [inline ecl/eclcc/eclcc.cpp:174]: if sa_handler is SIG_DFL then sa_flags could probably be set to 0.

File: ecl/hql/hqlrepository.cpp [hunk 3]
```diff
@@ -26,6 +26,36 @@
 #include "hqlutil.hpp"
 #include "jsecrets.hpp"
 
+static std::atomic<unsigned> fetchNesting;
+static std::atomic<bool> abortPending;
```
Comment (jakesmith) [inline ecl/hql/hqlrepository.cpp:30]: hesitated that these were auto-initialized to 0, clearer if explicitly {0} / {false} ?

**Review by jakesmith — APPROVED** [review]
@ghalliday - one minor comment.

**Review by mckellyln — COMMENTED** [review]
A few comments.

File: ecl/hql/hqlrepository.cpp (hunk 3 above)
Comment (ghalliday) [reply ecl/hql/hqlrepository.cpp:30]: statics are always zero initialised, but no harm in making it explicit.

File: ecl/eclcc/eclcc.cpp [hunk 4]
```diff
@@ -126,16 +128,54 @@ void __cdecl IntHandler(int)
...
-
-MODULE_INIT(INIT_PRIORITY_STANDARD)
+static void installSignalHandlers()
 {
     signal(SIGINT, IntHandler);
-    return true;
+}
+
+static void restoreSignalHandlers()
+{
```
Comment (mckellyln) [inline ecl/eclcc/eclcc.cpp:137]: could maybe add signal(SIGINT, SIG_DFL);

File: ecl/eclcc/eclcc.cpp [hunk 5]
```diff
@@ -126,16 +128,54 @@ void __cdecl IntHandler(int)
...
+        _exit(2);
+    }
+}
+
+static void installSignalHandlers()
+{
+    struct sigaction act;
+    act.sa_sigaction = &sighandler;
+    sigemptyset(&act.sa_mask);
+    act.sa_flags = SA_SIGINFO; // set so sa_sigaction field is used
```
Comment (mckellyln) [inline ecl/eclcc/eclcc.cpp:164]: Could add SA_RESTART to sa_flags maybe to help any syscall that was interrupted in doMain().

**Review by mckellyln — APPROVED** [review]
Two comments, but approved as is.

**Pull request merged**